src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/MapperVisitor.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DotNet.ApiCompatibility.Abstractions
+{
+    /// <summary>
+    /// Class that implements a visitor pattern to visit the tree for a given mapper.
+    /// </summary>
+    public class MapperVisitor
+    {
+        /// <summary>
+        /// Visits the tree for the given <see cref="ElementMapper{T}"/>.
+        /// </summary>
+        /// <typeparam name="T">Underlying type for the objects that the mapper holds.</typeparam>
+        /// <param name="mapper"><see cref="ElementMapper{T}"/> to visit.</param>
+        public void Visit<T>(ElementMapper<T> mapper)
+        {
+            if (mapper is AssemblySetMapper assemblySetMapper)
+            {
+                Visit(assemblySetMapper);
+            }
+            else if (mapper is AssemblyMapper assemblyMapper)
+            {
+                Visit(assemblyMapper);
+            }
+            else if (mapper is NamespaceMapper nsMapper)
+            {
+                Visit(nsMapper);
+            }
+            else if (mapper is TypeMapper typeMapper)
+            {
+                Visit(typeMapper);
+            }
+            else if (mapper is MemberMapper memberMapper)
+            {
+                Visit(memberMapper);
+            }
+        }
+
+        /// <summary>
+        /// Visits the <see cref="AssemblySetMapper"/> and visits each <see cref="AssemblyMapper"/> in the mapper.
+        /// </summary>
+        /// <param name="mapper">The <see cref="AssemblySetMapper"/> to visit.</param>
+        public virtual void Visit(AssemblySetMapper mapper)
+        {
+            foreach (AssemblyMapper assembly in mapper.GetAssemblies())
```

Comment:
do we need to add a path for the mapper being null here ?

Reply:
Good catch.